def.cf

```diff
@@ -4,30 +4,95 @@
 #  - common/global variables and classes here
 #
 ###############################################################################
+body file control
+{
+      inputs => { @(def.extra_inputs) };
+}
 
 bundle common def
 
 {
+  classes:
+      # all extra classes are defined true
+      "$(extra_classes)" expression => "any", meta => { "extra" };
+
+      # all override classes are defined true
+      "$(override_classes)" expression => "any", meta => { "override" };
+
+      "have_augments_file" expression => fileexists($(augments_file)), scope => "bundle";
+
+      "augments_are_json" expression => regcmp(".+\.json", $(augments_file)), scope => "bundle";
+      "have_json_augments" and => { "have_augments_file", "augments_are_json" }, scope => "bundle";
+
+      "augments_are_yaml" expression => regcmp(".+\.yaml", $(augments_file)), scope => "bundle";
+      "have_yaml_augments" and => { "have_augments_file", "augments_are_yaml" }, scope => "bundle";
+
+      "have_roles" expression => isvariable("augments[roles]"), scope => "bundle";
+      "have_inputs" expression => isvariable("augments[inputs]"), scope => "bundle";
+
+    have_roles::
+      "$(extra_roles[byname][$(roles_byname_keys)])" expression => classmatch("$(roles_byname_keys)"), meta => { "role" };
+      "$(roles_byrole_keys)" expression => classmatch("$(extra_roles[byrole][$(roles_byrole_keys)])"), meta => { "role" };
 
   vars:
 
     any::
+      "augments_file" string => ifelse(fileexists("$(this.promise_dirname)/services/def.yaml"), "$(this.promise_dirname)/services/def.yaml",
```

Comment:
I thought it was tidier, because the other things in `services` are disposable, but I've move it now.